### ISSUE TITLE
Error: Undefined External 'WinMain' in module c0pe.nas

### ISSUE THREAD

**Issue by iahunq2** [LADSoft/OrangeC#971]
Very simple program:

```
#include <windows.h>

int WINAPI wWinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance,
               PWSTR szCmdLine, int CmdShow) {

    MessageBoxW(NULL, szCmdLine, L"Title", MB_OK);

    return 0;
}
```

Tried with both `occ /Wg` and `gccocc -mwindows`.

**Comment by LADSoft** [comment]
hm i'll have to work on this do we need a wmain() too?

**Comment by tqo50396** [comment]
Change the code to this and it will compile:

```
#include <windows.h>

int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance,
               LPSTR szCmdLine, int CmdShow) {

    MessageBoxW(NULL, szCmdLine, L"Title", MB_OK);

    return 0;
}
```

**Comment by GitMensch** [comment]
According to https://learn.microsoft.com/windows/win32/learnwin32/winmain--the-application-entry-point the only thing to add is looking for _either_ of both, preferably `wWinain` - not sure how to achieve this, though.

**Comment by LADSoft** [comment]
we don't quite do linking the same way as microsoft...   so...   without a rewrite this isn't likely to happen without additional compiler switches.   I may consider a rewrite later, not right now though....

**Comment by Serge3leo** [comment]
Workaround:
```c
#include <processenv.h>

int WINAPI wWinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance,
                    PWSTR pCmdLine, int nCmdShow);

int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance,
                   PSTR /*lpCmdLine*/, int nCmdShow)
{
    return wWinMain(hInstance, hPrevInstance, GetCommandLineW(), nCmdShow);
}
```
It has a small drawback: an unnecessary conversion from Unicode to local encoding is performed for WinMain(), but the results are not used.

@LADSoft 
But if you put `wWinMain()` or `WinMain()` in library (or declare as weak), it will work automatically without any additional compiler switches. That's how I see it.

**Comment by LADSoft** [comment]
the easy thing to do here is to add an additional compiler switch for character set...   it would default to ASCII (VC calls it no character set) but could be set to use wide characters.   The current scheme also has problems because if you define both main and winmain and/or dll main you get multiply defined symbols...    Really what needs to happen here is that the linker needs to figure out which startup routine to call....   there would also be a change to the runtime libraries.   not a lot of work I guess....

**Comment by Serge3leo** [comment]
As far as I know, neither MSVC nor Pelles C have additional compiler switches, but they provide the necessary selection through the link tools. Of course, you know better how to do this in Orange C, but an additional switches would be an inconvenient and unique solution.

**Comment by LADSoft** [comment]
the thing is, there has to be some way of selecting the code that calls main() or wmain()...   once you have that code you can use the normal external selection to select either main or winmain but can't select that code without knowing what is being called...   which is not a 'normal' function for a linker.    but I guess I could have the linker look explicitly for those symbols to select an appropriate library function to do the calling :smile:

**Comment by GitMensch** [comment]
Exactly + possibly provide an option for the selection (= enforcement on the command line which makes the auto-resolving unneeded)... but if you're sure the linker check always works correctly and doesn't take much time even in case of many linker input, then that could be left out.

**Comment by LADSoft** [comment]
yeah i thought about this today, we already have switches for main vs winmain vs dllmain so it is just a matter of on the main and winmain cases doing a search for a couple of globals,  then in the main and winmain cases we would search the linked objects for the two character set options and choose one.    Since technically we can do incremental links with this linker I'd have too check for the presence of the objects before including them lol...   but I wouldn't think it would be a performance issue.

I'm going to put the function bindings in the linker definition file just to keep specific function names out of the linker source code :smile:

**Comment by LADSoft** [comment]
so i made good progress on this...   i rewrote it and mostly it works the way i want, resolves the various problems nicely.   Cleans up the directory tree nicely too, to not have all those c0xxx.nas files lieing around :smile:.   Right now I'm trying to get LSCRTL.DLL to work again...   should be done this week sometime...

**Comment by Serge3leo** [comment]
> the easy thing to do here is to add an additional compiler switch for character set...

A small note about the way `WinMain()` (`wWinMain()`) is searched. Roughly speaking, the current state of affairs is:
```
if (CMAKE_C_COMPILER_ID STREQUAL OrangeC)
  list(APPEND my_sources "occ_workaround.c")
elseif (MINGW AND NOT MINGW32CE)
  add_link_options(-municode)  # clang/gcc need for use `wWinMain()`
endif ()
# MSVC and Pelles C don't require any action
```
Personally, I think the "do nothing" option is the 21st century style, but if the switch matches the `clang/gcc` switch, it's still an improvement.

**Comment by LADSoft** [comment]
I'm not adding a switch for charset but you will still have to set /DUNICODE manually to the compiler command line so the compiler knows which prototypes to use when compiling your source files...   this is pretty much what visual studio does when you select the character set as unicode i believe...

switches exist for console/gui/dll already though because the linker needs to know which subsystem to generate an executable for...

the state of this issue is that I've done basic testing and gotten it compiling the runtime library, but there are some issues with compiling the test projects that I'm ironing out.
